File: 프로그래머스/2/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.py
import heapq

def solution(sequence, k):
    answer = []
    dp = [sequence[0]]
    length = len(sequence)
    
    start = 0 
    end = 0
    
    pq = []
    
    for i in range(1,length):
        dp.append(sequence[i]+dp[i-1])        
    
    while start <= end and end < length:

        
        if  start == 0:
            s = dp[end]
        else:
            s = dp[end]-dp[start-1]

 
        if s == k:
            if not pq or pq[0][0] > end-start+1:
                heapq.heappush(pq,(end-start+1,[start,end]))
            end+=1
        elif s > k:
            start+=1
        else:
            end+=1
    
    
    return heapq.heappop(pq)[1]

File: 프로그래머스/2/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.js
function solution(sequence, k) {
    let answer = [0,sequence.length-1];
    
    let start = 0;
    let end = 0;
    let sum = sequence[end];
    
    while(start <= end && end < sequence.length){
 
        if(sum === k){
            if(answer[1]-answer[0] > end-start){
                answer = [start,end]
            }
            sum-=sequence[start]
            start+=1 
        }
        else if(sum < k){
            end+=1
            sum+=sequence[end]
        }
        else{
            sum-=sequence[start]
            start+=1
           
        }
        
    }
    
    return answer;
}